Add MongoDB Connection

CathrinTruchan opened the pull request:
Hello :) 
I added a mongoDB connection to my app. The PR is quite large because I had to adjust many parts of the code. 
Looking forward to your feedback! 
Thanks a lot!

Kheirah commented on models/Flow.js:
```diff
@@ -0,0 +1,18 @@
+});
+
+const Flow =
+  mongoose.models.Flow || mongoose.model("Flow", flowSchema, "flows");
```
Same applies here. Consider renaming the model name to lowercase letters in case I'm not wrong about my assumption of mongoose converting everything to lowercase.

Kheirah commented on pages/api/flows/index.js:
```diff
@@ -0,0 +1,46 @@
+    const flows = await getAllFlows();
+    return response.status(200).json(flows);
+  } else if (request.method === "POST") {
+    await dbConnect();
+
+    const postData = JSON.parse(request.body);
+    const newFlow = await Flow.create(postData);
+
+    return response
+      .status(201)
+      .json({ message: "Flow created", createdId: newFlow.id });
+  } else if (request.method === "PATCH") {
+    await dbConnect();
```
Consider calling `dbConnect()` once at the very top of your handler function.

CathrinTruchan replied:
Done

Kheirah commented on pages/api/flows/index.js:
```diff
@@ -0,0 +1,46 @@
+  } else if (request.method === "PATCH") {
+    await dbConnect();
+    const updatedData = JSON.parse(request.body);
+    console.log(updatedData);
```
You may remove logs which are not essential to a well-functioning app.

Kheirah commented on pages/api/flows/index.js:
```diff
@@ -0,0 +1,46 @@
+      .status(200)
+      .json({ message: "Flow updated", name: updatedFlow.name });
+  } else if (request.method === "DELETE") {
+    await dbConnect();
```
Same here, call dbConnect only once at the top.

Kheirah commented on pages/flow/[id].js:
```diff
@@ -1,47 +1,61 @@
+      });
+
+      const result = await response.json();
+      console.log(result.name);
```
Remove unused logs

Kheirah commented on services/asanaService.js:
```diff
@@ -0,0 +1,41 @@
+
+export async function getAsanaById(id) {
+  await dbConnect();
+  const asanas = await Asana.find();
+  const asana = asanas.find((asana) => asana.id === id);
```
There's an inbuilt function in mongoose called `findById` which you can use to directly pass down the id. But be aware, you are returned an object as a result, not an array.

CathrinTruchan replied:
Done

Kheirah commented on services/flowService.js:
```diff
@@ -0,0 +1,36 @@
+
+export async function getFlowById(id) {
+  await dbConnect();
+  const flows = await Flow.find();
+  const flow = flows.find((flow) => flow.id === id);
```
Same applies here, you can use `findById`.

CathrinTruchan replied:
Done

Kheirah reviewed (changes requested):
Overall really good work!

The changes I'm proposing are all minor ones. But since they are so numerous, I won't immediately approve the PR and instead wait until you had the chance to make adjustments. When you're done, let me know, and I'll give you that checkmark :)

Or, if I'm unavailable, ask someone else for an approval after your changes.

CathrinTruchan commented on components/CreateFlowForm.js:
```diff
@@ -3,28 +3,40 @@ import styled from "styled-components";
     const name = data.name.trim();
 
+    const flowData = {
+      id: id,
+      name: name,
```
Done :)

CathrinTruchan commented on pages/asanas/[id].js:
```diff
@@ -14,7 +16,8 @@ export async function getStaticPaths() {
 export async function getStaticProps(context) {
   const { id } = context.params;
-  const asana = getAsanaByID(id);
+  console.log(id);
```
Done

CathrinTruchan commented on services/asanaService.js:
```diff
@@ -0,0 +1,41 @@
+  await dbConnect();
+
+  const asanas = await Asana.find();
+  console.log(asanas);
```
Done

CathrinTruchan commented on services/flowService.js:
```diff
@@ -0,0 +1,36 @@
+  await dbConnect();
+
+  const flows = await Flow.find();
+  console.log(flows);
+  console.log("test");
```
Done

CathrinTruchan commented:
> Overall really good work!
> 
> The changes I'm proposing are all minor ones. But since they are so numerous, I won't immediately approve the PR and instead wait until you had the chance to make adjustments. When you're done, let me know, and I'll give you that checkmark :)
> 
> Or, if I'm unavailable, ask someone else for an approval after your changes.

Hello @Kheirah, 
thank you so much for your detailed review and feedback. I have implemented all your changes. I am just not sure about the model naming issue? I added a comment below. 
And sorry for the many console.logs. I have been so much back and forth in this PR that I haven't noticed them anymore 😅

Kheirah commented on models/Asana.js:
```diff
@@ -0,0 +1,19 @@
+});
+
+const Asana =
+  mongoose.models.Asana || mongoose.model("Asana", asanaSchema, "asanas");
```
You're right. Then it's all fine. I had an issue myself the other day where I had a model called "tlaRental" and mongoose converted all letters to lowercase ones, but my collection in the database was called "tlaRental". Mongoose couldn't find the collection because it was looking for a collection in all lowercase. That's why I mentioned this lowercase thing. But if it's working for you, all good :)

Kheirah reviewed (approved):
Very good!
